Feature/sc 32779/search module route updates for clark refactor

Romerico234 opened the pull request:
Removed and updated GET_PUBLIC_LEARNING_OBJECTS_WITH_FILTER route to point to search.service.ts from featured.service.ts and learning-object.service.ts

Cwagne17 reviewed (commented):
Can you rebase to the branch you worked off of and then re-request.

pzalep1 commented on src/app/core/learning-object-module/search/search.service.ts:
```diff
@@ -1,9 +1,94 @@
+   * @returns {Promise<LearningObject[]>}
+   * @memberof LearningObjectService
+   */
+  getLearningObjects(
```
Same thing as the other PRs. If you change the class that the function is in all of the references will need to be updated. VS Code has some shortcuts that make it easy to find where the references are.

pzalep1 reviewed (commented):
Looks good! Just need to update the references to the functions that you moved.

Romerico234 commented:
Working on fixing function calls

Romerico234 commented:
There was another search.service.ts in the admin directory so I removed it and moved its functions to search.service.ts in the  learning-object-module. And I updated the function calls.

pzalep1 commented on src/app/core/feature-module/featured.service.ts:
```diff
@@ -148,34 +145,6 @@ export class FeaturedObjectsService {
     }
   }
 
-  /**
-   * Fetches Array of Learning Objects that are not currently featured
-   *
-   * @returns {Promise<LearningObject[]>}
-   * @memberof LearningObjectService
-   */
-  getNotFeaturedLearningObjects(
-    query?: Query,
-  ): Promise<{ learningObjects: LearningObject[]; total: number }> {
-    let route = '';
-    if (query) {
-      const queryClone = Object.assign({}, query);
-      const queryString = querystring.stringify(queryClone);
-      route =
-        SEARCH_ROUTES.GET_PUBLIC_LEARNING_OBJECTS_WITH_FILTER(queryString);
-    } else {
-      route = SEARCH_ROUTES.GET_PUBLIC_LEARNING_OBJECTS;
-    }
-
-    return this.http
-      .get(route)
-      .pipe(catchError(this.handleError))
-      .toPromise()
-      .then((response: any) => {
-        return { learningObjects: response.objects, total: response.total };
-      });
-  }
-
```
I believe this should still be located here or the admin dashboard will break.

Cwagne17 replied:
This is actually using a learning-objects search route. I would update the component that uses it and it should be okay. I just think the name of the function should be updated to match how we refer to it in the backend.

Cwagne17 commented on src/app/core/learning-object-module/search/search.service.ts:
```diff
@@ -1,9 +1,100 @@
+      }
+      const queryString = new URLSearchParams(queryClone).toString();
+      route =
+        SEARCH_ROUTES.GET_PUBLIC_LEARNING_OBJECTS_WITH_FILTER(queryString);
```
can we rename the search route to match the name we use in the backend? `searchLearningObjects` and then condense these functions into one?

Cwagne17 commented on src/app/core/learning-object-module/search/search.service.ts:
```diff
@@ -1,9 +1,100 @@
 })
 export class SearchService {
+  needsChange$: Subject<void> = new Subject();
 
```
This route needs to be moved into the search module also. This comes from the learning object search module route handler in clark-service. 

https://github.com/Cyber4All/clark-service/blob/main/src/modules/learning-object-module/search/search.routehandler.ts

![Screenshot 2024-09-16 at 10 09 34 AM](https://github.com/user-attachments/assets/148ac766-ad1c-4894-a132-b5c22bcc10de)

Romerico234 commented:
I combined the two functions into a single function, searchLearningObjects. To distinguish between the functionalities of the original functions within this combined function, I added optional parameters that handle standard outcomes and the mapping to LearningObject instances. These optional parameters are set to true by default.

Additionally, I updated the function calls to this new function to specify the appropriate options. Specifically, for any component that previously used getNotPublicLearningObjects (which was only one), I set both options to false to replicate its original behavior.

johngonz0 commented on src/app/admin/pages/learning-objects/learning-objects.component.ts:
```diff
@@ -18,11 +17,12 @@ import { ToastrOvenService } from '../../../shared/modules/toaster/notification.
   templateUrl: './learning-objects.component.html',
   styleUrls: ['./learning-objects.component.scss'],
-  providers: [PublicLearningObjectService]
+  providers: []
```
does there need to be anything in this providers array?

pzalep1 replied:
Not with the new organizational structure we are doing with the services.

johngonz0 reviewed (commented):
just a question

Cwagne17 commented on src/app/core/learning-object-module/search/search.routes.ts:
```diff
@@ -1,8 +1,16 @@
 export const SEARCH_ROUTES = {
-  GET_PUBLIC_LEARNING_OBJECTS: `${environment.apiURL}/learning-objects`,
-  GET_PUBLIC_LEARNING_OBJECTS_WITH_FILTER(query: string) {
+  needsChange$,
+  REGISTER_CHANGE() {
+    this.needsChange$.next();
+  },
```
This should not be here, only routes to our APIs should be in the routes.ts files.

Romerico234 replied:
I'm a little confused with what you mean by this. Are you asking me to move this route into the search.routehandler.ts file in clark-service?

Cwagne17 commented on src/app/core/learning-object-module/search/search.routes.ts:
```diff
@@ -1,8 +1,16 @@
+    this.needsChange$.next();
+  },
+
+  SEARCH_LEARNING_OBJECTS: `${environment.apiURL}/learning-objects`,
+  SEARCH_LEARNING_OBJECTS_WITH_FILTER(query: string) {
```
These can be combined.

Cwagne17 commented on src/app/core/learning-object-module/search/search.service.ts:
```diff
@@ -1,9 +1,90 @@
+    query?: Query,
+    options?: {
+      handleStandardOutcomes?: boolean;
+      mapToLearningObjectInstance?: boolean;
```
When would we not want to map to a learning object class?

Cwagne17 commented on src/app/core/learning-object-module/search/search.service.ts:
```diff
@@ -1,9 +1,90 @@
+  registerChange() {
+    SEARCH_ROUTES.REGISTER_CHANGE();
+  }
+
```
We are still missing getUserLearningObjects function.

Romerico234 commented:
Waiting for Chris' PR to get merged then we'll see what to do with this PR.

Romerico234 commented:
I moved registerChange route back to the search.service.ts in admin/core directory. I combined the function and variable in search.route.ts. Lastly, I pulled updates from clark-service-refactor.
